To configure username and password using Helm

baba230896 opened the pull request:
## Summary

Added implementation to configure the username and password for YSQL and YCQL using Helm.

**Flow:**
Added post-install hook job to set up the credentials. The post-install hook job gets the script from ConfigMap. The hook job sets the custom credentials in the environment vars & the `setup credentials` script uses those environment vars for configuration.

[EDITED]
**Note:** 

- The changes work with TLS and Non-TLS clusters.

~~- The changes only work with the Non-TLS cluster.~~
~~- Compatible Changes for TLS is in-progress.~~

## Test Plan:
The changes have been tested with the mentioned env vars combination in `yugabyted` [reference docs](https://docs.yugabyte.com/latest/reference/configuration/yugabyted/#environment-variables). 

The following commands to test the changes (These changes are from the demo branch, and the branch is in development)-

1. The following commands are used to spin up the cluster with the custom `YSQL` password.
```sh
helm repo remove yugabytedb
helm repo add yugabytedb https://baba230896.github.io/charts
helm repo update
helm search repo yugabytedb/yugabyte
helm install yb-demo yugabytedb/yugabyte \
    --set resource.master.requests.cpu=150m,resource.master.requests.memory=100Mi,storage.master.size=3Gi \
    --set resource.tserver.requests.cpu=150m,resource.tserver.requests.memory=100Mi,storage.tserver.size=3Gi \
    --set replicas.tserver=3,replicas.master=3 \
    --set authCredentials.ysql.password=sqlpassword \
    --namespace ys-helm 
```
2. Verify the new password (Provide the new password in prompt).
```sh
kubectl exec -n ys-helm -it yb-tserver-0 -- ysqlsh -h yb-tserver-0.yb-tservers.ys-helm
```

## Fixes:
https://github.com/yugabyte/yugabyte-db/issues/5663

ramkumarvs commented on stable/yugaware/templates/configs.yaml:
```diff
@@ -191,6 +191,24 @@ data:
 
+      - job_name: 'kube-state-metrics'
+        static_configs:
+        - targets: ['kube-state-metrics.kube-system.svc.cluster.local:8080']
```
what if the cluster domain name is different?

Arnav15 replied:
Addressed comment.

ramkumarvs commented on stable/yugaware/templates/configs.yaml:
```diff
@@ -191,6 +191,24 @@ data:
             target_label: "saved_name"
             replacement: "$1"
 
+      - job_name: 'kube-state-metrics'
```
do we need any changes to our RBAC permissions to query kube state metrics?

Arnav15 replied:
No, that can be setup when bringing up the kube_state_metrics service.

baba230896 committed:
To configure username and password using Helm

Added post-install hook job to set up the credentials.
The post-install hook job get the script from ConfigMap.
The hook job set the custom credentials in environment
vars and script use those environment vars for config-
-uration.

For more env vars combination,
https://docs.yugabyte.com/latest/reference/configuration/yugabyted/#environment-variables

baba230896 committed:
To configure username and password via helm values in TLS secure cluster

iSignal commented on stable/yugabyte/templates/hooks/setup-credentials-job.yaml:
```diff
@@ -0,0 +1,80 @@
+        {{- end }}
+        command:
+        - 'bash'
+        - '/home/yugabyte/bin/setup-credentials/setup-credentials.sh'
```
does the return value of this cmd matter?

baba230896 replied:
Yeah because of this only container will know the script is executed successfully or not.

iSignal replied:
Makes sense - but it looks like when this cmd fails, the entire release is considered failed? Can you try that scenario to make sure that the error is easy to diagnose?

baba230896 replied:
In this case, the debugging is easy because, in the end, it gives the error something like this 'post-hook failed'. We're removing the post-install hook job's pod only on success so, in case of failure, we can check the logs of the job and can do further debugging.

iSignal commented on stable/yugabyte/templates/setup-credentials-configmap.yaml:
```diff
@@ -0,0 +1,259 @@
+
+      echo "${ysql_cmd[@]}"
+      while ! "${ysql_cmd[@]}"; do
+        sleep 5s
```
would be good to log something like "Waiting for YSQL interface to become available". 

Was thinking that if the cluster never becomes healthy, this hook will never complete and the pod will never become Ready/Running. Will kubectl describe be able to tell us what is happening in that case?

baba230896 replied:
log is already added at line no. 217.

This part exists in the `waitUntilHealthy` function and, this function is used with a timeout of 3m at line no.218. The script will wait for 3 mins to get the cluster healthy, but somehow the cluster is unhealthy then the script exit with status 1. And the "backoffLimit" is 2 so K8s will execute the hook two times in case of failure and even after 2 executions hook failed then the `kubectl/helm` exit and return the error something like "post hook failed". We're removing the post-install hook job pod only on success so in case of failure, we can check the logs of the job and can do some debugging.

baba230896 committed:
Commented the tserver gflags part in the YB values.yaml
